src/main/java/org/commcare/cases/util/StorageBackedTreeRoot.java

```diff
@@ -291,23 +297,41 @@ protected void initBasicQueryHandlers(QueryPlanner queryPlanner) {
         }
 
 
-        String[] namesToMatch = new String[numKeysToProcess];
-        String[] valuesToMatch = new String[numKeysToProcess];
+        Vector<String> namesToMatch = new Vector<>();
+        Vector<String> valuesToMatch = new Vector<>();
+        Vector<String> namesToInverseMatch = new Vector<>();
+        Vector<String> valuesToInverseMatch = new Vector<>();
 
         String cacheKey = "";
         String keyDescription ="";
 
         for (int i = numKeysToProcess - 1; i >= 0; i--) {
-            namesToMatch[i] = profiles.elementAt(i).getKey();
-            valuesToMatch[i] = (String)
-                    (((IndexedValueLookup)profiles.elementAt(i)).value);
-
-            cacheKey += "|" + namesToMatch[i] + "=" + valuesToMatch[i];
-            keyDescription += namesToMatch[i] + "|";
+            String name;
+            String value;
+            String operator;
+            if (profiles.elementAt(i) instanceof IndexedValueLookup) {
+                name = profiles.elementAt(i).getKey();
+                value = (String)(((IndexedValueLookup)profiles.elementAt(i)).value);
+                namesToMatch.add(name);
+                valuesToMatch.add(value);
+                operator = "=";
+            } else {
+                name = profiles.elementAt(i).getKey();
+                value = (String)(((NegativeIndexedValueLookup)profiles.elementAt(i)).value);
+                namesToInverseMatch.add(name);
+                valuesToInverseMatch.add(value);
+                operator = "!=";                
+            }
+            cacheKey += "|" + name + operator + value;
+            keyDescription += name + "|";
         }
+        String[] namesArray = namesToMatch.toArray(new String[namesToMatch.size()]);
+        String[] valuesArray = valuesToMatch.toArray(new String[valuesToMatch.size()]);
+        String[] inverseNames = namesToInverseMatch.toArray(new String[namesToInverseMatch.size()]);
+        String[] inverseValues = valuesToInverseMatch.toArray(new String[valuesToInverseMatch.size()]);
         mMostRecentBatchFetch = new String[2][];
-        mMostRecentBatchFetch[0] = namesToMatch;
-        mMostRecentBatchFetch[1] = valuesToMatch;
+        mMostRecentBatchFetch[0] = namesArray;
```

Comment:
You'll need to make sure the NegativeIndexedValue lookups make it into this key set as well. This is what is used to index the cache of results, so if the negative indexes aren't in there, if you first query `[a = b][c != d]` then query `[a=b]` it'll give you the cached result from the first query.

Reply:
Where is this used? I could not actually find the usages of this field, just a few places where it is set.

Reply:
This may only be used in platform specific code. android appears to have a request path for it that goes through the following:
https://github.com/dimagi/commcare-android/blob/f1f383e478850de92418e8f1057422b3ecd84595/app/src/org/commcare/engine/cases/AndroidCaseInstanceTreeElement.java#L237

https://github.com/dimagi/commcare-android/blob/f1f383e478850de92418e8f1057422b3ecd84595/app/src/org/commcare/models/AsyncNodeEntityFactory.java#L103

I believe it's also used in Child Index resolution, but would need to look deeper through the call stack